### PR TITLE
Enable new win-arm64 platform

### PR DESCRIPTION
**Pull request by gaborkertesz-linaro**
This PR enables the new Windows on Arm (win-arm64) platform, by implementing the required APIs, reading topology information via Windows API.

### CODE REVIEW

**Comment by gaborkertesz-linaro** [comment]
[_win-arm64_cmake-config.log](https://github.com/pytorch/cpuinfo/files/8233034/_win-arm64_cmake-config.log)
[_win-arm64_cmake-build.log](https://github.com/pytorch/cpuinfo/files/8233038/_win-arm64_cmake-build.log)
[_win-arm64_init-test.log](https://github.com/pytorch/cpuinfo/files/8233039/_win-arm64_init-test.log)
[_win-arm64_cache-info.log](https://github.com/pytorch/cpuinfo/files/8233041/_win-arm64_cache-info.log)
[_win-arm64_cpu-info.log](https://github.com/pytorch/cpuinfo/files/8233042/_win-arm64_cpu-info.log)
[_win-arm64_isa-info.log](https://github.com/pytorch/cpuinfo/files/8233044/_win-arm64_isa-info.log)

**Comment by gaborkertesz-linaro** [comment]
Build on win-arm64
`cmake . -G "Visual Studio 16 2019" -A ARM64
cmake --build .`

File: src/arm/windows/init.c [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+	/* 2. Read vendor/manufacturer name from registry. */
+	if (!read_registry(cpu0_subkey, vendor_name_value, &textBuffer)) {
+		cpuinfo_log_error("Registry read error");
+		goto cleanup;
+	}
+
+	for (uint32_t i = 0; i < (sizeof(vendors) / sizeof(struct vendor_info)); i++) {
+		if (strncmp(textBuffer, vendors[i].vendor_name,
+				strlen(vendors[i].vendor_name)) == 0) {
+			*vendor = vendors[i].cpu_info_vendor;
```
Comment (marcpems) [inline src/arm/windows/init.c:206]: same - needs a break here, right?

File: src/arm/windows/init.c [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+	}
+	return result;
+}
+
+static void get_isa()
+{
+	bool armv8 = IsProcessorFeaturePresent(PF_ARM_V8_INSTRUCTIONS_AVAILABLE);
+	bool crypto = IsProcessorFeaturePresent(PF_ARM_V8_CRYPTO_INSTRUCTIONS_AVAILABLE);
+	bool load_store_atomic = IsProcessorFeaturePresent(PF_ARM_64BIT_LOADSTORE_ATOMIC);
+	bool divide = IsProcessorFeaturePresent(PF_ARM_DIVIDE_INSTRUCTION_AVAILABLE);
```
Comment (marcpems) [inline src/arm/windows/init.c:226]: there are a bunch of processor feature flags that you retrieve and are unused besides the debug log. 
Can you remove everything that is not used or move it into a #ifdef debug block with the log line.

File: src/arm/windows/init_by_logical_sys_info.c [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+	HANDLE heap = GetProcessHeap();
+
+	/* 1. Query the size of the information structure first */
+	if (GetLogicalProcessorInformationEx(info_type, NULL, &info_size) == FALSE) {
+		const DWORD last_error = GetLastError();
+		if (last_error != ERROR_INSUFFICIENT_BUFFER) {
+			cpuinfo_log_error(
+				"failed to query size of processor %"PRIu32" information information: error %"PRIu32"",
+				(uint32_t)info_type, (uint32_t) last_error);
+			return FALSE;
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:455]: minor - should be a 0 not FALSE. Or maybe throw exception?

File: src/arm/windows/init_by_logical_sys_info.c [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+	if (infos == NULL) {
+		cpuinfo_log_error("failed to allocate %"PRIu32" bytes for logical processor information",
+			(uint32_t) info_size);
+		return 0;
+	}
+	/* 3. Read the information structure */
+	if (GetLogicalProcessorInformationEx(info_type, infos, &info_size) == FALSE) {
+		cpuinfo_log_error("failed to query processor %"PRIu32" information: error %"PRIu32"",
+			(uint32_t)info_type, (uint32_t) GetLastError());
+		HeapFree(heap, 0, infos);
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:469]: add a cleanup: label and use a goto, for consistent memory clean up behaviour.

File: src/arm/windows/init_by_logical_sys_info.c [hunk 5]
```diff
@@ -0,0 +1,842 @@
...
+		KAFFINITY group_processors_mask = info->Processor.GroupMask[i].Mask;
+		while (group_processors_mask != 0) {
+			const uint32_t processor_id_in_group =
+				low_index_from_kaffinity(group_processors_mask);
+			const uint32_t processor_global_index =
+				global_proc_index_per_group[group_id] + processor_id_in_group;
+
+			if(processor_global_index >= nr_of_processors) {
+				cpuinfo_log_error("unexpected processor index %"PRIu32"",
+					processor_global_index);
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:555]: this shouldnt happen, but if it does then the store_package... calls are likely to throw

File: src/arm/windows/init_by_logical_sys_info.c [hunk 6]
```diff
@@ -0,0 +1,842 @@
...
+						cores, chip_info);
+				break;
+				default:
+					cpuinfo_log_error(
+						"unexpected processor info type (%"PRIu32") for processor information",
+						(uint32_t) info->Relationship);
+				break;
+			}
+			/* Reset the lowest bit in affinity mask */
+			group_processors_mask &= (group_processors_mask - 1);
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:578]: This resets the most significant bits, not the lowest bit.

File: src/arm/windows/init_by_logical_sys_info.c [hunk 7]
```diff
@@ -0,0 +1,842 @@
...
+		cores[core_id].core_id = core_id;
+		get_core_uarch_for_efficiency(
+			chip_info->chip_name, core_info->Processor.EfficiencyClass,
+			&(cores[core_id].uarch), &(cores[core_id].frequency));
+
+		/* We don't have cluster information, so we handle it as
+		 * fixed 1 to (cluster / cores).
+		 */
+		processors[processor_global_index].cluster =
+			(const struct cpuinfo_cluster*) NULL + core_id;
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:728]: I dont understand what this is doing or how its safe. cluster as a pointer is being set to an absolute address equivalent to the core_id * sizeof(cpuinfo_cluster). Did I miss something?

File: src/arm/windows/init.c [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+	result &= cpu_info_init_by_logical_sys_info(chip_info, vendor);
+	cpuinfo_is_initialized = result;
+	return result;
+}
+
+bool get_core_uarch_for_efficiency(
+	enum woa_chip_name chip, BYTE EfficiencyClass,
+	enum cpuinfo_uarch* uarch, uint64_t* frequency)
+{
+	if (!uarch || !frequency) {
```
Comment (marcpems) [inline src/arm/windows/init.c:82]: this function could be simplified and remove the switch - test chip for < woa_chip_name_last and just assign based on chip. 
if (uarch && frequency && EfficiencyClass < 2)
{
*uarch = woa_chips[chip].uarchs[EfficiencyClass].uarch;
*frequency = woa_chips[chip].uarchs[EfficiencyClass].frequency;
return true;
}

File: src/arm/windows/init_by_logical_sys_info.c [hunk 9]
```diff
@@ -0,0 +1,842 @@
...
+	processors = HeapAlloc(heap, HEAP_ZERO_MEMORY, nr_of_processors * sizeof(struct cpuinfo_processor));
+	if (processors == NULL) {
+		cpuinfo_log_error(
+			"failed to allocate %zu bytes for descriptions of %"PRIu32" logical processors",
+			nr_of_processors * sizeof(struct cpuinfo_processor), nr_of_processors);
+		goto clean_up;
+	}
+
+	/* 2. Read topology information via MSDN API: packages, cores and caches*/
+	nr_of_packages = read_packages_for_processors(
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:170]: there are many things that can go wrong in these calls, but there is no way to chain the error and exit the function on error. That's not great for defensive code.
Consider adding throw/catch if permitted.

File: src/arm/windows/init_by_logical_sys_info.c [hunk 10]
```diff
@@ -0,0 +1,842 @@
...
+	processors = NULL;
+	packages = NULL;
+	clusters = NULL;
+	cores = NULL;
+	caches = NULL;
+	uarchs = NULL;
+
+clean_up:
+	if (global_proc_index_per_group != NULL) {
+		HeapFree(heap, 0, global_proc_index_per_group);
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:329]: its safe to call HeapFree() with a null ptr for the 3rd param. 
Drop the NULL check and just call it.
This is a style thing, but I would always zero the variable after calling free just to show its not usable past that point.

File: src/arm/windows/init_by_logical_sys_info.c [hunk 11]
```diff
@@ -0,0 +1,842 @@
...
+	if (caches == NULL) {
+		cpuinfo_log_error("failed to allocate %zu bytes for descriptions of %"PRIu32" caches",
+			nr_of_all_caches * sizeof(struct cpuinfo_cache), nr_of_all_caches);
+		goto clean_up;
+	}
+
+	/* 4.Read missing topology information that can't be saved without counted
+	 *   allocate structures in the first round.
+	 */
+	read_caches_for_processors(
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:246]: no error checking here. Exception handling would solve that.

File: src/arm/windows/init_by_logical_sys_info.c [hunk 12]
```diff
@@ -0,0 +1,842 @@
...
+		struct cpuinfo_processor* processor = processors + processor_id;
+
+		struct cpuinfo_core* core = (struct cpuinfo_core*)processor->core;
+
+		/* We stored the offset of pointers when we haven't allocated memory
+		 * for packages and clusters, so now add offsets to base addresses.
+		 */
+		struct cpuinfo_package* package =
+			(struct cpuinfo_package*) ((uintptr_t) packages + (uintptr_t) processor->package);
+		processor->package = package;
```
Comment (marcpems) [inline src/arm/windows/init_by_logical_sys_info.c:786]: this needs a sanity check, i.e. processor->package should not be less than it's original value (overflow) and not be greater than the total size of packages.

**Review by marcpems — COMMENTED** [review]
The general principals of querying CPU details and assigning to the cpuinfo relevant structures looks ok to me. There are some issues with the error handling here that might need addressing.

File: src/arm/windows/init_by_logical_sys_info.c (hunk 7 above)
Comment (marcpems) [reply src/arm/windows/init_by_logical_sys_info.c:728]: I figured it out from the code later that patches these values by adding a base offset. Can you add a comment here to make it clear?

File: src/arm/windows/init.c (hunk 8 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init.c:82]: That's absolutely true, I agree. However I wanted to create a pattern for to help adding future platform by others by emphasizeing that the connection of micro-architecture information and EfficiencyClass can be various, depending on the given SoC.
Would you still prefer to simplify the code?

File: src/arm/windows/init.c [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+	if (!read_registry(cpu0_subkey, chip_name_value, &textBuffer)) {
+		cpuinfo_log_error("Registry read error");
+		goto cleanup;
+	}
+	for (uint32_t i = 0; i < (uint32_t) woa_chip_name_last; i++) {
+		size_t compare_length = strnlen(woa_chips[i].chip_name_string, CPUINFO_PACKAGE_NAME_MAX);
+		int compare_result = strncmp(textBuffer, woa_chips[i].chip_name_string, compare_length);
+		if (compare_result == 0) {
+			*chip_info = woa_chips+i;
+		}
```
[the remark this answers is not in the document]
Comment (gaborkertesz-linaro) [reply src/arm/windows/init.c:189]: Absolutely, will be fixed!

File: src/arm/windows/init.c (hunk 1 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init.c:206]: Yes, thanks for the catch!

File: src/arm/windows/init_by_logical_sys_info.c [hunk 14]
```diff
@@ -0,0 +1,842 @@
...
+			nr_of_cores * sizeof(struct cpuinfo_core), nr_of_cores);
+		goto clean_up;
+	}
+
+	/* We allocate one contiguous cache array for all caches, then use offsets per cache type. */
+	uint32_t counted_sum_caches = 0;
+	for (uint32_t i = 0; i < MAX_NR_OF_CACHES; i++) {
+		counted_sum_caches += numbers_of_caches[i];
+	}
+	if (counted_sum_caches != nr_of_all_caches) {
```
[the remark this answers is not in the document]
Comment (gaborkertesz-linaro) [reply src/arm/windows/init_by_logical_sys_info.c:230]: I wanted to add sanity check, but this can be removed.

File: src/arm/windows/init_by_logical_sys_info.c (hunk 6 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init_by_logical_sys_info.c:578]: I've inherited this code from x86 init: https://github.com/pytorch/cpuinfo/blob/master/src/x86/windows/init.c#L229
I mean it's absolutely not an excuse, I'll check that again.

File: src/arm/windows/init_by_logical_sys_info.c (hunk 7 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init_by_logical_sys_info.c:728]: Again, I've copied this pattern from x86: https://github.com/pytorch/cpuinfo/blob/master/src/x86/windows/init.c#L224
I've struggled a lot to understand why it's useful and I've tried to add as many as clear as I could, but as you've struggled also, I've failed with that!

File: src/arm/windows/init_by_logical_sys_info.c (hunk 9 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init_by_logical_sys_info.c:170]: As far as I see, there are not throw/catch in the current codebase and I'd think that's intentional, so I'd try to cleanup the error paths and make sure it's going to be safe.

File: src/arm/windows/init_by_logical_sys_info.c (hunk 6 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init_by_logical_sys_info.c:578]: I think it really clears all bits, that are lower than the least set bit.

`ff	11111111
fe	11111110
fc	11111100
f8	11111000
f0	11110000
e0	11100000
c0	11000000
80	10000000
0	00000000`

`0x10
0x0

0x20
0x0

0x40
0x0

0x80
0x0`

File: src/arm/windows/init.c [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+	}
+	if (*chip_info == NULL) {
+		cpuinfo_log_error("Unknown chip model name.\n Please add new Windows on Arm SoC/chip support!");
```
[the remark this answers is not in the document]
Comment (gaborkertesz-linaro) [reply src/arm/windows/init.c:192]: SoC type per core (called uarch) is not stored per core. We only get the EfficiencyClass per core, currently 0 or 1, where 0=little core, 1=big core.
Setting the frequency is optional.

File: src/arm/windows/init_by_logical_sys_info.c (hunk 9 above)
Comment (nsait-linaro) [reply src/arm/windows/init_by_logical_sys_info.c:170]: Yes `C` code mostly

**Comment by gaborkertesz-linaro** [comment]
@malfet Could you review it? Thanks!

**Comment by gaborkertesz-linaro** [comment]
related PR:
https://github.com/pytorch/pytorch/pull/72424

File: cmake/DownloadGoogleBenchmark.cmake [hunk 16]
```diff
@@ -4,8 +4,8 @@ PROJECT(googlebenchmark-download NONE)
 
 INCLUDE(ExternalProject)
 ExternalProject_Add(googlebenchmark
-	URL https://github.com/google/benchmark/archive/v1.2.0.zip
-	URL_HASH SHA256=cc463b28cb3701a35c0855fbcefb75b29068443f1952b64dd5f4f669272e95ea
+	URL https://github.com/google/benchmark/archive/v1.6.1.zip
+	URL_HASH SHA256=367e963b8620080aff8c831e24751852cffd1f74ea40f25d9cc1b667a9dd5e45
```
Comment (malfet) [inline cmake/DownloadGoogleBenchmark.cmake:8]: Can you please do it as separate PR? (As it is completely unrelated to change in question, is it?)

File: src/arm/windows/init.c [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+		}
+	}
+};
+
+BOOL CALLBACK cpuinfo_arm_windows_init(
+	PINIT_ONCE init_once, PVOID parameter, PVOID* context)
+{
+	struct woa_chip_info *chip_info = NULL;
+	enum cpu_info_vendor vendor = cpuinfo_vendor_unknown;
+	BOOL result = false;
```
Comment (malfet) [inline src/arm/windows/init.c:71]: C99 supports `bool` type, isn't it?
```suggestion
	bool result = false;

```

File: src/arm/windows/windows_arm_init.h [hunk 18]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+bool get_core_uarch_for_efficiency(
+	enum woa_chip_name chip, BYTE EfficiencyClass,
+	enum cpuinfo_uarch* uarch, uint64_t* frequency);
+
+BOOL cpu_info_init_by_logical_sys_info(
```
Comment (malfet) [inline src/arm/windows/windows_arm_init.h:7]:
```suggestion
bool cpu_info_init_by_logical_sys_info(

```

File: src/arm/windows/init_by_logical_sys_info.c [hunk 19]
```diff
@@ -0,0 +1,877 @@
...
+	struct cpuinfo_cluster* clusters,
+	struct cpuinfo_core* cores,
+	const uint32_t nr_of_cores,
+	const struct woa_chip_info* chip_info,
+	enum cpuinfo_vendor vendor);
+
+static inline uint32_t low_index_from_kaffinity(KAFFINITY kaffinity);
+
+
+BOOL cpu_info_init_by_logical_sys_info(
```
Comment (malfet) [inline src/arm/windows/init_by_logical_sys_info.c:119]:
```suggestion
bool cpu_info_init_by_logical_sys_info(
```

**Review by malfet — APPROVED** [review]
Mostly looks good to me, but please explain why `ARM64` needs a special treatment in `ELSEIF(NOT CPUINFO_TARGET_PROCESSOR_MATCHES` 
And use `bool` instead of `BOOL` whenever possible (i.e. in callback its probably fine, but otherwise should not be used)

File: cmake/DownloadGoogleBenchmark.cmake (hunk 16 above)
Comment (gaborkertesz-linaro) [reply cmake/DownloadGoogleBenchmark.cmake:8]: win-arm64 needs this patch to pass benchmark build:
https://github.com/google/benchmark/commit/ea5a5bbff491fd625c6e3458f6edd680b8bd5452

`benchmark.lib(sysinfo.obj) : error LNK2019: unresolved external symbol __rdtsc referenced in function "__int64 __cdecl benchmark::cycleclock::Now(void)" (?Now@cycleclock@benchmar
k@@YA_JXZ) [C:\kg\python_test\github_packages\cpuinfo\init-bench.vcxproj]
C:\kg\python_test\github_packages\cpuinfo\Debug\init-bench.exe : fatal error LNK1120: 1 unresolved externals [C:\kg\python_test\github_packages\cpuinfo\init-bench.vcxproj]`

So if you still would prefer I can create a new PR with this change, I just wanted to give you the background.

File: CMakeLists.txt [hunk 20]
```diff
@@ -65,7 +65,7 @@ IF(NOT CMAKE_SYSTEM_PROCESSOR)
       "cpuinfo will compile, but cpuinfo_initialize() will always fail.")
     SET(CPUINFO_SUPPORTED_PLATFORM FALSE)
   ENDIF()
-ELSEIF(NOT CPUINFO_TARGET_PROCESSOR MATCHES "^(i[3-6]86|AMD64|x86(_64)?|armv[5-8].*|aarch64|arm64)$")
+ELSEIF(NOT CPUINFO_TARGET_PROCESSOR MATCHES "^(i[3-6]86|AMD64|x86(_64)?|armv[5-8].*|aarch64|arm64)$|ARM64")
```
[the remark this answers is not in the document]
Comment (gaborkertesz-linaro) [reply CMakeLists.txt]: Good point, no good reason, fixed!

**Comment by gaborkertesz-linaro** [comment]
@malfet Can I do anything else to close this and the related PR?

**Comment by strega-nil-ms** [comment]
any updates on this?

**Comment by gaborkertesz-linaro** [comment]
> any updates on this?

Not on my side, but I'd happily close this!

**Comment by nsait-linaro** [comment]
Can this be merged, please?

**Comment by kalaskarsanket** [comment]
Any plans on merging this PR ?

File: src/arm/windows/init.c [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+/* Please add new vendor here! */
+static struct vendor_info vendors[] = {
+	{
+		"Qualcomm",
+		cpuinfo_vendor_qualcomm
+	}
+};
+
+/* Please add new SoC/chip info here! */
+static struct woa_chip_info woa_chips[] = {
```
Comment (Maratyszcza) [inline src/arm/windows/init.c:37]: This covers literally two devices. We need a more generic detection method that doesn't require hard-coding every SoC.

File: CMakeLists.txt [hunk 22]
```diff
@@ -171,6 +171,9 @@ IF(CPUINFO_SUPPORTED_PLATFORM)
       LIST(APPEND CPUINFO_SRCS
         src/arm/android/properties.c)
     ENDIF()
+    ELSEIF(CMAKE_SYSTEM_NAME STREQUAL "Windows" AND CPUINFO_TARGET_PROCESSOR STREQUAL "ARM64")
```
Comment (Maratyszcza) [inline CMakeLists.txt:174]: Wrong indentation?

**Review by Maratyszcza — CHANGES_REQUESTED** [review]
Use dash `-` to separate filename parts rather than underscore `_` for consistency with other files in cpuinfo.

File: src/arm/windows/init.c [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+/* Efficiency class = 0 means little core, while 1 means big core for now */
+#define MAX_WOA_VALID_EFFICIENCY_CLASSES		2
+#define VENDOR_NAME_MAX		CPUINFO_PACKAGE_NAME_MAX
+
+struct cpuinfo_arm_isa cpuinfo_isa;
+
+static void set_cpuinfo_isa_fields();
+static bool get_system_info_from_registry(
+	struct woa_chip_info** chip_info,
+	enum cpu_info_vendor* vendor);
```
Comment (kaadam) [inline src/arm/windows/init.c:21]: typo: cpu_info_vendor -> cpuinfo_vendor

File: src/arm/windows/init_by_logical_sys_info.c [hunk 24]
```diff
@@ -0,0 +1,877 @@
...
+
+	switch(info->Cache.Level) {
+		case 1:
+			switch (info->Cache.Type) {
+				case CacheInstruction:
+					processors[processor_global_index].cache.l1i = current_cache;
+				break;
+				case CacheData:
+					processors[processor_global_index].cache.l1d = current_cache;
+				break;
```
Comment (kaadam) [inline src/arm/windows/init_by_logical_sys_info.c:775]: same here as above

File: src/arm/windows/init_by_logical_sys_info.c [hunk 25]
```diff
@@ -0,0 +1,877 @@
...
+	if (caches == NULL) {
+		switch(info->Cache.Level) {
+			case 1:
+				switch (info->Cache.Type) {
+					case CacheInstruction:
+						numbers_of_caches[cpuinfo_cache_level_1i]++;
+					break;
+					case CacheData:
+						numbers_of_caches[cpuinfo_cache_level_1d]++;
+					break;
```
Comment (kaadam) [inline src/arm/windows/init_by_logical_sys_info.c:627]: Clang-cl complains about the missing other two enum values:
"error: enumeration values 'CacheUnified' and 'CacheTrace' not handled in switch [-Werror,-Wswitch]"
Could you add these values as empty cases as well?

File: src/arm/windows/init.c [hunk 26]
```diff
@@ -0,0 +1,252 @@
...
+	{
+		"Qualcomm",
+		cpuinfo_vendor_qualcomm
+	}
+};
+
+/* Please add new SoC/chip info here! */
+static struct woa_chip_info woa_chips[] = {
+	/* Microsoft SQ1 Kryo 495 4 + 4 cores (3 GHz + 1.80 GHz) */
+	"Microsoft SQ1",
```
Comment (kaadam) [inline src/arm/windows/init.c:39]: Coud you add some extra braces to each struct elements? 
../../third_party/cpuinfo/src/src/arm/windows/init.c(39,2): error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]                                                                                                         "Microsoft SQ1",                                                                                                        ^~~~~~~~~~~~~~~~                                                                                                        {                                                                                                               ../../third_party/cpuinfo/src/src/arm/windows/init.c(52,2): error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]                                                                                                         "Microsoft SQ2",                                                                                                        ^~~~~~~~~~~~~~~~                                                                                                        {

**Comment by nsait-linaro** [comment]
(@gaborkertesz-linaro is on holiday now and planning to address the comments in couple of weeks)

File: src/arm/windows/init.c (hunk 21 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init.c:37]: I absolutely agree that this solution is not ideal, but we haven't found any better and still robust option considering the existing Windows API. Also new SoCs won't turn up frequently, so even though it's a limited solution, it can be maintained.

**Comment by gaborkertesz-linaro** [comment]
I hope I've fixed all of your findings, please let me know if I missed any of them!
Unfortunetaly I wasn't able to setup building with LLVM by CMake, but I've built with '-Wall' by MSVC compiler.

File: src/arm/windows/init.c (hunk 21 above)
Comment (nsait-linaro) [reply src/arm/windows/init.c:37]: I had a similar concern but there doesn't seem to be any better way to do it with the help of Windows API.

Ampere SoC also runs Windows/Arm64 - Does that need to be added?

File: src/arm/windows/init-by-logical-sys-info.c [hunk 27]
```diff
@@ -0,0 +1,879 @@
...
+	}
+	current_cache->processor_count++;
+
+	switch(info->Cache.Level) {
+		case 1:
+			switch (info->Cache.Type) {
+				case CacheInstruction:
+					processors[processor_global_index].cache.l1i = current_cache;
+				break;
+				case CacheData:
```
Comment (kaadam) [inline src/arm/windows/init-by-logical-sys-info.c:775]: Just a nit: Could you add 'CacheUnified' and 'CacheTrace' to this switch as well?

File: src/arm/windows/init.c (hunk 21 above)
Comment (gaborkertesz-linaro) [reply src/arm/windows/init.c:37]: Would be nice, but I can't really test that, so I'd ask to do as a follow-up commit.

**Review by kaadam — COMMENTED** [review]
Great, thanks. I'm not a reviewer, but I give you an informal LGTM.

**Comment by gaborkertesz-linaro** [comment]
@malfet What do you think?

**Review by malfet — APPROVED** [review]
